frontend/src/environments/dotenv-loader.js

```diff
@@ -0,0 +1,16 @@
+const dotenv = require('dotenv');
+
+/**
+ * @see {https://github.com/Lacka90/angular-dotenv-poc/blob/master/src/app/config/dotenv-loader.js}
+ */
+
+dotenv.config();
+
+const env = {
+  mapsKey: process.env.MAPS_KEY || '',
+  functionsKey: process.env.FUNCTIONS_KEY || '',
+};
+
+module.exports = () => {
+  return { code: 'module.exports = ' + JSON.stringify(env) };
```

Comment:
It seems quite complex, I've never had store those values as json and then use some additional loaders, like val-loader just to use dontenv values. It's something that Angular architecture requires?

Reply:
Unfortunately, after my reasearch it seems that Angular CLI and its automated build system makes it hard to do it in an easy way (which is weird and seems like a huge omission from Angular team, but maybe they did it purposefully).

I've found 2 approaches that work for our case:
1. Overriding the default builders in angular.json and using 3rd-party custom builders that claim full backwards compatibility with default builders. They extend default builders with custom webpack config possibilities on top. See: https://medium.com/@fidelisclayton/system-environment-variables-in-angular-1f4a922c7b4c

However, I decided against this approach, as there's a risk that this 3rd party builders will not be maintained in the future if we'd like to migrate to newer Angular versions etc.

2. Using inline val-loader supported by webpack, as this does not require to change the default builders, but brings some overhead of additional typings for typescript and "proxy" loader files. See the source article: https://codingsans.com/blog/configure-frontend-projects-with-dotenv

And related webpack docs for loaders:
https://webpack.js.org/concepts/loaders/#inline
https://webpack.js.org/loaders/val-loader/

If you think there's a better approach, please let me know :)

Reply:
OK, thanks for the explanation. Makes sense now. I also think the 2nd option is better. Good call 👍 
I was using in the past just simple DefinePlugin or EnvironmentPlugin
(I see there's webpack-dotenv-plugin now), but all of them require access to Webpack config.

https://webpack.js.org/plugins/environment-plugin/
https://github.com/nwinch/webpack-dotenv-plugin